test/cpp/end2end/client_lb_end2end_test.cc

```diff
@@ -240,6 +241,7 @@ class ClientLbEnd2endTest : public ::testing::Test {
     servers_.clear();
     creds_.reset();
     grpc_shutdown_blocking();
+    grpc_wait_for_async_shutdown();
```

Comment:
Do we know where is the pending async shutdown from in this case?
Since there is a comment right above trying to manually drop all the grpc_init refcount, I think it is useful to fully understand this issue.

Reply:
I can look into that, but I'm not sure if that matters a lot though. It looks like there's not a way for a user or a test to "wait for grpc clean-up done" in gRPC core's semantics. I think we need to either add a function to provide this semantic (as in the commit above) or change the test to not depend on this semantic.

Reply:
So far the pending async shutdown I see is always in [`destroy_channel()`](https://github.com/grpc/grpc/blob/23c8cfcfdaeb1191eeb0e0823b1badb6be83d815/src/core/lib/surface/channel.cc#L520). Typically the stack looks like this:
```
 grpc_shutdown
 destroy_channel()
 exec_ctx_run()
 grpc_core::ExecCtx::Flush()
 grpc_core::Executor::RunClosures()
 grpc_core::Executor::ThreadMain()
 grpc_core::(anonymous namespace)::ThreadInternalsPosix::ThreadInternalsPosix()::{lambda()#1}::operator()()
 grpc_core::(anonymous namespace)::ThreadInternalsPosix::ThreadInternalsPosix()::{lambda()#1}::__invoke()
 start_thread
```

Probably due to a call reference that is not removed in the test's main thread (which kind of makes sense to me, but I'm not familiar with it enough to say if anything is broken or not).

Reply:
Looks like `destroy_channel()` can be called in an executor: https://github.com/grpc/grpc/blob/master/src/core/lib/transport/transport.cc#L53. That can explain why `grpc_shutdown` is not called immediately after clearing up things on the user side.

Reply:
Thanks. That makes sense. I think what happens is you have multiple servers queuing grpc_shutdown into executor and when the main thread is doing shutdown_blocking, those grpc_shutdown calls have not been all executed.

I would suggest you remove the grpc_init and shutdown_blocking in the test because they do not really do what you need. To wait for the final shutdown, there are several ways.
1. Remove the maybe_wait function and use your function to replace all call sites.
2. roll a simple version in your test such as while (grpc_is_initialized()) { sleep(some time); } and then call maybe_wait_for_async_shutdown.

Reply:
I tend to avoid depending on the shutdown in the test at all. See the updated commit.